symforce/slam/imu_preintegration/imu_factor.h

```diff
@@ -0,0 +1,192 @@
+
+#include <vector>
+
+#include <Eigen/Core>
+
+#include <sym/pose3.h>
+#include <symforce/opt/factor.h>
+#include <symforce/opt/key.h>
+
+#include "./preintegrated_imu_measurements.h"
+
+namespace sym {
+
+/**
+ * A factor for using on-manifold IMU preintegration in a SymForce optimization problem.
+ *
+ * Assumes IMU bias is constant over the preintegration window. Does not recompute the
+ * preintegrated measurements for the IMU bias arguments, but rather a first order
+ * approximation linearized at the the IMU biases given during preintegration.
+ *
+ * The gravity argument should be [0, 0, -g] (where g is 9.8, assuming your world frame is
+ * gravity aligned so that the -z direction points towards the Earth) unless your IMU reads 0
+ * acceleration while stationary, in which case it should be [0, 0, 0].
+ *
+ * More generally, the gravity argument should yield the true acceleration when added to the
+ * accelerometer measurement (after the measurement has been converted to the world frame).
+ *
+ * Is a functor so as to store the preintegrated IMU measurements between two times.
+ * The residual computed is the local-coordinate difference between the final state (pose
+ * and velocity) and the state you would expect given the initial state and the IMU measurements.
+ *
+ * Based heavily on the on manifold ImuFactor found in GTSAM and on the paper:
+ *
+ * Christian Forster, Luca Carlone, Frank Dellaert, and Davide Scaramuzza,
+ * “IMU Preintegration on Manifold for Efficient Visual-Inertial Maximum-a-Posteriori Estimation”,
+ * Robotics: Science and Systems (RSS), 2015.
+ *
+ * Example Usage:
+ *
+ *   enum Var : char {
+ *     POSE = 'p',        // Pose3d
+ *     VELOCITY = 'v',    // Vector3d
+ *     ACCEL_BIAS = 'A',  // Vector3d
+ *     GYRO_BIAS = 'G',   // Vector3d
+ *     GRAVITY = 'g',     // Vector3d
+ *     EPSILON = 'e'      // Scalar
+ *   };
+ *
+ *   struct ImuMeasurement {
+ *     Eigen::Vector3d acceleration;
+ *     Eigen::Vector3d angular_velocity;
+ *     double duration;
+ *   };
+ *
+ *   int main() {
+ *     // Dummy function declarations
+ *     std::vector<ImuMeasurement> GetMeasurementsBetween(double start_time, double end_time);
+ *     std::vector<double> GetKeyFrameTimes();
+ *     Eigen::Vector3d GetAccelBiasEstimate(double time);
+ *     Eigen::Vector3d GetGyroBiasEstimate(double time);
+ *     void AppendOtherFactors(std::vector<sym::Factord> & factors);
+ *
+ *     // Example accelerometer and gyroscope covariances
+ *     const Eigen::Matrix3d accel_cov = 1e-5 * Eigen::Matrix3d::Identity();
+ *     const Eigen::Matrix3d gyro_cov = 1e-5 * Eigen::Matrix3d::Identity();
+ *
+ *     std::vector<sym::Factord> factors;
+ *
+ *     // GetKeyFrameTimes assumed to return at least 2 times
+ *     std::vector<double> key_frame_times = GetKeyFrameTimes();
+ *
+ *     // Integrating Imu measurements between keyframes, creating an ImuFactor for each
+ *     for (unsigned int i = 0; i < key_frame_times.size() - 1; i++) {
```

Comment:
Elsewhere we use the convention `int i = 0; i < static_cast<int>(key_frame_times.size()) - 1; i++`, although it doesn't really matter for example code - if we're going to use something that's unsigned / not `int` though we should probably be consistent, and it should probably be `size_t`